### PR TITLE
[REVIEW] duration types

### PR DESCRIPTION
**Pull request by karthikeyann**
addresses C++ part of https://github.com/rapidsai/cudf/issues/5272
- [x] add duration types
- [x] factories

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by kkraus14** [comment]
@karthikeyann Could we limit the scope of this PR to just the types and factories and use follow up PRs for other pieces? This also lets us parallelize the work more efficiently 😄

File: cpp/include/cudf/types.hpp [hunk 1]
```diff
@@ -191,6 +195,16 @@ enum type_id {
   TIMESTAMP_MILLISECONDS,  ///< duration of milliseconds since Unix Epoch in int64
   TIMESTAMP_MICROSECONDS,  ///< duration of microseconds since Unix Epoch in int64
   TIMESTAMP_NANOSECONDS,   ///< duration of nanoseconds since Unix Epoch in int64
+  DURATION_YEARS,          ///< duration of years in int32
```
[the remark this answers is not in the document]
Comment (kkraus14) [reply cpp/include/cudf/types.hpp:198]: We need days since we have `TIMESTAMP_DAYS`, but otherwise agreed for now.

In the future the goal would be to support most of these though 😢

[the remark this answers is not in the document]
Comment (karthikeyann) [reply cpp/include/cudf/types.hpp:198]: YEARS/MONTHS/WEEKS/DAYS/MINUTES would be useful for constructing all timedelta operations. This would allow  for eg. (+1 year + 1 month +1 days) operations with timestamp. 
This is still open for discussion.

[the remark this answers is not in the document]
Comment (jrhemstad) [reply cpp/include/cudf/types.hpp:198]: > useful for constructing all timedelta operations

Useful, yes, practical, no. I do not think our compile time or library file size can handle add all of these types in addition to all of the other new ones being added.

File: cpp/src/dictionary/set_keys.cu [hunk 2]
```diff
@@ -85,7 +85,6 @@ struct dispatch_compute_indices {
              cudaStream_t stream)
   {
     CUDF_FAIL("list_view dictionary set_keys not supported yet");
-    return nullptr;
```
Comment (jrhemstad) [inline cpp/src/dictionary/set_keys.cu:88]: I suspect some compiler versions are going to complain about not having a return value here. Furthermore, this is an unrelated change.

File: cpp/include/cudf/types.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (kkraus14) [reply cpp/include/cudf/types.hpp:198]: FWIW, Arrow only supports durations on seconds, milliseconds, microseconds, and nanoseconds.

Numpy supports it on years, months, weeks, days, hour, and minutes as well but uses 64 bit types for all of them.

Maybe we only support seconds, milliseconds, microseconds, and nanoseconds for now, considering using a duration of 64 bit seconds allows you to capture all of a duration of 32 bit days, albeit with memory / performance tradeoffs.

[the remark this answers is not in the document]
Comment (jrhemstad) [reply cpp/include/cudf/types.hpp:198]: I'm definitely in favor of supporting the minimal set of essential types.

File: cpp/src/dictionary/set_keys.cu (hunk 2 above)
Comment (karthikeyann) [reply cpp/src/dictionary/set_keys.cu:88]: IIRC, compiler complains if return type is auto.

File: cpp/include/cudf/types.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (revans2) [reply cpp/include/cudf/types.hpp]: For spark we need MONTHS, DAYS, and MICROSECONDS.

This is because if I want to go one year in the future it is 12 months, but it is not always 365 days.  If I want to go 1 week in the future it is always 7 days, but I cannot always say how many microseconds it is.

File: cpp/include/cudf/wrappers/durations.hpp [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+/**
+ * @file durations.hpp
+ * @brief Concrete type definitions for int32_t and int64_t durations in varying resolutions.
+ **/
+namespace cudf {
+namespace detail {
+
+template <class Duration>
+struct duration : Duration {
```
[the remark this answers is not in the document]
Comment (jrhemstad) [reply cpp/include/cudf/wrappers/durations.hpp:34]: The duration helper types like `std::chrono::days` don't provide sufficient guarantees for the size of the type, e.g., `std::chrono::days` is only guaranteed to be at least 25 bits. 

That said, I can't recall why we made the explicit `timestamp` type instead of doing:
```
using timestamp_D = simt::std::chrono::time_point<simt::std::chrono::system_clock, simt::std::chrono::duration<int32_t, simt::std::ratio<86400>>>
```

@trxcllnt do you remember why we didn't make the libcudf timestamp types simply be aliases and instead made a concrete new type? I'm sure we had a good reason, but I can't remember right now.

File: cpp/include/cudf/types.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (kkraus14) [reply cpp/include/cudf/types.hpp]: Well, looks like `std::chrono::duration` is all just ratio based: https://en.cppreference.com/w/cpp/chrono/duration

| Unit | Ratio |
| - | - |
| seconds | 1 |
| minutes | 60 |
| hours | 3600 |
| days | 86400 |
| weeks | 604800 |
| months | 2629746 |
| years | 31556952 |

[the remark this answers is not in the document]
Comment (jrhemstad) [reply cpp/include/cudf/types.hpp]: > This is because if I want to go one year in the future it is 12 months, but it is not always 365 days.

According to C++ duration types, a year is always 365.2425 days. See:https://en.cppreference.com/w/cpp/chrono/duration

>  years is equal to 365.2425 days (the average length of a Gregorian year). months is equal to 30.436875 days (exactly 1/12 of years).

File: cpp/include/cudf/wrappers/durations.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (trxcllnt) [reply cpp/include/cudf/wrappers/durations.hpp:34]: Yeah I think there were a few things:

1. We didn't want to make the clock part of the type, since we expected to switch to `utc_clock` eventually once it lands in libcu++.
2. We have an extra constructor that takes a numeric `typename Duration::rep r` type that's more convenient to use
3. For some reason we needed to override the conversion constructor to explicitly call `duration_cast()` since the [libcu++ impl](https://github.com/rapidsai/thirdparty-libcxx/blob/a97a7380c76346c22bb67b93695bed19592afad2/include/chrono#L1383) doesn't call it (possible bug in libcu++?).

[the remark this answers is not in the document]
Comment (sriramch) [reply cpp/include/cudf/wrappers/durations.hpp:34]: for the 2nd item above - should we be even supporting constructing a timestamp with a tick count? in this case, what does 1000 really represent? shouldn't it be a duration (of ms in this example) or another time_point with a similar or a different duration. presently, we interpret to mean add 1000 ms to the system_clock epoch (as we only support system_clock for now), which doesn't look to be the original intent, as 1000 could mean different units or displacement from some other clock.

i think 3rd item is required if we want to truncate and lose precision - i.e. assigning say `timestamp_s` to a `timestamp_D`. that constructor will prevent this if the periods aren't divisible - `is_convertible` should evaluate to true.

[the remark this answers is not in the document]
Comment (karthikeyann) [reply cpp/include/cudf/wrappers/durations.hpp:34]: 2). 2nd item is present for ease the testing code. `timestamp(typename Duration::rep r)` should be **explicit**. 
3) 3rd item is not bug since `is_convertible` takes care of it. (lossy conversion is not allowed as per `std::chrono`, needs explicit `duration_cast`). `std::chrono` does not recommend using `duration_cast` (lossy conversion) implicitly. One plus point is that it's explicit constructor.
Perhaps lossy conversion is the requirement from users (Spark, Python, etc). is it?

[the remark this answers is not in the document]
Comment (jrhemstad) [reply cpp/include/cudf/wrappers/durations.hpp:34]: @karthikeyann do you think we should move away from having our own concrete `timestamp` type and instead just us a type alias for both timestamps and durations?

[the remark this answers is not in the document]
Comment (karthikeyann) [reply cpp/include/cudf/wrappers/durations.hpp:34]: We can't replace due to a CUDA limitation. If `timestamp` is replaced with `simt::std::chrono::time_point`, following error is hit due to limitation of shared memory initialization. 
`cpp/include/cudf/detail/copy_if.cuh(106): error: initializer not allowed for __shared__ variable `
https://github.com/rapidsai/cudf/blob/efebcd1452692ee15f3d30627a9ef3d0cafa85d5/cpp/include/cudf/detail/copy_if.cuh#L106

This is a limitation of device variable. https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#device-memory-specifiers
- The constructor function has no parameters, the initializer list is empty and the function body is an empty compound statement.
- The default constructors of all base classes of its class can be considered empty.

`simt::std::chrono::time_point` calls parameterized constructor of _duration_
https://github.com/rapidsai/thirdparty-libcxx/blob/a97a7380c76346c22bb67b93695bed19592afad2/include/chrono#L1372

`    _LIBCPP_INLINE_VISIBILITY _LIBCPP_CONSTEXPR_AFTER_CXX11 time_point() : __d_(duration::zero()) {}`

`timestamp` class avoids it by calling no parameter constructor of _duration_.

File: cpp/include/cudf/scalar/scalar.hpp [hunk 4]
```diff
@@ -416,4 +416,76 @@ class timestamp_scalar : public detail::fixed_width_scalar<T> {
...
+/**
+ * @brief An owning class to represent a duration value in device memory
+ *
+ * @ingroup scalar_classes
+ *
+ * @tparam T the data type of the duration value
+ * @see cudf/wrappers/durations.hpp for a list of allowed types
+ */
+template <typename T>
+class duration_scalar : public detail::fixed_width_scalar<T> {
```
[the remark this answers is not in the document]
Comment (karthikeyann) [reply cpp/include/cudf/scalar/scalar.hpp]: modified and added.

File: cpp/include/cudf/types.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (karthikeyann) [reply cpp/include/cudf/types.hpp]: now, DAYS, SECONDS, MILLISECONDS, MICROSECONDS, NANOSECONDS are the only types in duration type.

File: cpp/include/cudf/wrappers/durations.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (karthikeyann) [reply cpp/include/cudf/wrappers/durations.hpp]: `simt::std::chrono::duration` is used for duration types. 
for timestamp, wait for #5400 .

File: cpp/include/cudf/wrappers/durations.hpp [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+namespace cudf {
+
+/**
+ * @addtogroup duration_classes Duration
+ * @{
+ */
+/**
+ * @brief Type alias representing an int32_t duration of days.
+ **/
+using duration_D = simt::std::chrono::duration<int32_t, simt::std::chrono::days::period>;
```
Comment (jrhemstad) [inline cpp/include/cudf/wrappers/durations.hpp:39]: Can we use these as the duration type of the timestamp types?

File: cpp/include/cudf/wrappers/timestamps.hpp [hunk 6]
```diff
@@ -45,6 +45,7 @@ struct timestamp : time_point<Duration> {
    *
    * @param other The `timestamp` to copy
    */
+  // TODO: This explict truncation is intended?
```
Comment (sriramch) [inline cpp/include/cudf/wrappers/timestamps.hpp:48]: would this one *and* the constructor `constexpr timestamp(Duration d)` have to be explicitly defined? they are:

- either relaxing or adding explicit specifiers that aren't in the chrono's `time_point` type
- allows narrowing/truncation of types, when it is disallowed by `std::chrono`

is it not possible to inherit them from the base type?

**Comment by devavret** [comment]
Does this PR ensure that `timestamp_s - timestamp_s == duration_s`? Is that automatic or would this need overloading `timestamp_s::operator-()`

**Comment by sriramch** [comment]
> Does this PR ensure that `timestamp_s - timestamp_s == duration_s`? Is that automatic or would this need overloading `timestamp_s::operator-()`

the support for these operators (add/sub etc.) with the timestamp/duration operands will be added as a part of a follow-up pr's focusing on unary/binary ops.

File: cpp/include/cudf/wrappers/timestamps.hpp (hunk 6 above)
Comment (trxcllnt) [reply cpp/include/cudf/wrappers/timestamps.hpp:48]: Yeah the list of timestamp constructors evolved over time, and in response to various compiler errors (@karthikeyann reminds me of one I forgot in [this comment](https://github.com/rapidsai/cudf/pull/5359#discussion_r435774513)). Some of them may not be valid reasons any longer.

I'd say change/remove whichever ones we want, and recompile to see what breaks (this can be frustrating, as changing this header triggers a near full re-compile).

File: cpp/src/io/csv/csv_gpu.cu [hunk 7]
```diff
@@ -351,6 +353,23 @@ __inline__ __device__ cudf::timestamp_ns decode_value(const char *data,
   return milli * 1000000;
 }
 
+// The purpose of this is merely to allow compilation ONLY
+// TODO : make this work for json
```
Comment (devavret) [inline cpp/src/io/csv/csv_gpu.cu:357]: Is this a TODO to add implementation to properly support it or is it a reminder to do the same thing for json?

File: cpp/tests/utilities/column_utilities.cu [hunk 8]
```diff
@@ -336,6 +336,12 @@ struct column_view_printer {
     }
   }
 
+  template <typename Element, typename std::enable_if_t<is_duration<Element>()>* = nullptr>
+  void operator()(cudf::column_view const& col, std::vector<std::string>& out)
+  {
+    CUDF_FAIL("duration printing not supported yet");
```
[the remark this answers is not in the document]
Comment (devavret) [reply cpp/tests/utilities/column_utilities.cu:342]: Supposedly coming in a later PR where testing code will be updated to work with duration types.
